web_monitoring/tests/test_diffing_server_exc_handling.py

```diff
@@ -52,6 +52,29 @@ def test_both_local(self):
                 self.assertEqual(response.code, 200)
 
 
+class DiffingServerEtagTest(DiffingServerTestCase):
+    def test_etag_header(self):
+        with tempfile.NamedTemporaryFile() as a:
+            with tempfile.NamedTemporaryFile() as b:
+                cold_response = self.fetch('/html_token?format=json&include=all&'
+                        f'a=file://{a.name}&b=file://{b.name}')
+                self.assertEqual(cold_response.code, 200)
+
+                etag = cold_response.headers.get('Etag')
+
+                warm_response = self.fetch('/html_token?format=json&include=all&'
+                        f'a=file://{a.name}&b=file://{b.name}',
+                                           headers={'If-None-Match': etag,
+                                           'Accept': 'application/json'})
+                self.assertEqual(warm_response.code, 304)
+
+                mismatch_response = self.fetch('/html_token?format=json&include=all&'
+                        f'a=file://{a.name}&b=file://{b.name}',
+                                           headers={'If-None-Match': 'Stale Value',
+                                           'Accept': 'application/json'})
+                self.assertEqual(mismatch_response.code, 200)
```

Comment:
We should also have a test to make sure no Etag header is sent if there was an error on the server, e.g. if the content of the file can’t be decoded, or the content of `b` doesn’t match `b_hash`, or the URL of `b` is unreachable, etc. I don’t know that we need to exhaustively test all these cases, but we should test at least one.

Reply:
I added an assertion for the absence of the etag header in our tests where we assert a 4xx response code for diffs. Tornado appears to be doing the right thing for us since we're using its methods for computing + checking etags and sending errors.

Reply:
> Tornado appears to be doing the right thing for us since we're using its methods for computing + checking etags and sending errors.

It’s doing the right thing, but not because of that! *All* headers that have been set so far get thrown out when you call `send_error()` (which also gets called under the hood if you throw an exception). It’s more that this is good to have in the tests so that we know if we erroneously return error information by explicitly setting status code and calling `write()`.